File: Haskell/insertionsort/insertionsort_250000.hs
insertionsort :: Ord a => [a] -> [a]
insertionsort [] = []
insertionsort [x] = [x]
insertionsort (x:xs) = insert x (insertionsort xs)
  where
    insert x [] = [x]
    insert x (y:ys)
        | x <= y    = x : y : ys
        | otherwise = y : insert x ys


main = (insertionsort arr) `seq` (return ())

File: Rust/insertionsort/insertionsort_250000.rs
fn insertion_sort<T: Ord>(arr: &mut [T]) {
    for i in 1..arr.len() {
        let mut j = i;
        while j > 0 && arr[j] < arr[j - 1] {
            arr.swap(j, j - 1);
            j -= 1;
        }
    }
}

fn main() {
    let debug: bool = false;
    insertion_sort(&mut arr);
    if debug {
        println!("{:?}", arr);
    }    
}